File: platform/contracts/model-tiers.ts
import { z } from "zod";

export const ModelTierSchema = z.enum(["local", "mid", "frontier"]);
export const ModelTierFloorSchema = z.enum(["any", "local", "mid", "frontier"]);

export type ModelTier = z.infer<typeof ModelTierSchema>;
export type ModelTierFloor = z.infer<typeof ModelTierFloorSchema>;

export type RegisteredModelTier = {
  provider: string;
  model: string;
  tier: ModelTier;
  label: string;
  executable: boolean;
};

export const MODEL_TIERS = [
  {
    provider: "openai",
    model: "openai/gpt-5.2",
    tier: "frontier",
    label: "GPT-5.2",
    executable: true,
  },
  {
    provider: "openai",
    model: "openai/gpt-5.1-codex",
    tier: "frontier",
    label: "GPT-5.1 Codex",
    executable: true,
  },
  {
    provider: "openai",
    model: "openai/gpt-4.1-mini",
    tier: "mid",
    label: "GPT-4.1 Mini",
    executable: true,
  },
  {
    provider: "openai_codex",
    model: "openai_codex/gpt-5.3-codex",
    tier: "frontier",
    label: "GPT-5.3 Codex",
    executable: true,
  },
  {
    provider: "anthropic",
    model: "anthropic/claude-opus-4-6",
    tier: "frontier",
    label: "Claude Opus 4.6",
    executable: true,
  },
  {
    provider: "anthropic",
    model: "claude-opus-4-7",
    tier: "frontier",
    label: "Claude Opus 4.7",
    executable: true,
  },
  {
    provider: "anthropic",
    model: "anthropic/claude-sonnet-4-6",
    tier: "frontier",
    label: "Claude Sonnet 4.6",
    executable: true,
  },
  {
    provider: "anthropic",
    model: "anthropic/claude-haiku-4-5",
    tier: "mid",
    label: "Claude Haiku 4.5",
    executable: true,
  },
  {
    provider: "openai_compatible",
    model: "*",
    tier: "local",
    label: "OpenAI-compatible model",
    executable: true,
  },
  {
    provider: "local",
    model: "*",
    tier: "local",
    label: "Local relay model",
    executable: true,
  },
  {
    provider: "google",
    model: "google/gemini-2.5-pro",
    tier: "frontier",
    label: "Gemini 2.5 Pro",
    executable: false,
  },
  {
    provider: "mistral",
    model: "mistral/codestral-latest",
    tier: "mid",
    label: "Codestral",
    executable: false,
  },
] as const satisfies readonly RegisteredModelTier[];

export const MODEL_TIER_REGISTRY = MODEL_TIERS;

export type RegisteredProvider = (typeof MODEL_TIERS)[number]["provider"];

export function modelTier(
  provider: string | null | undefined,
  model: string | null | undefined,
): ModelTier | null {
  if (!provider || !model) return null;
  const exact = MODEL_TIERS.find(
    (entry) => entry.provider === provider && entry.model === model,
  );
  if (exact) return exact.tier;
  return (
    MODEL_TIERS.find(
      (entry) => entry.provider === provider && entry.model === "*",
    )?.tier ?? null
  );
}

export function modelRegistryEntry(
  provider: string | null | undefined,
  model: string | null | undefined,
): RegisteredModelTier | null {
  if (!provider || !model) return null;
  return (
    MODEL_TIERS.find(
      (entry) => entry.provider === provider && entry.model === model,
    ) ??
    MODEL_TIERS.find(
      (entry) => entry.provider === provider && entry.model === "*",
    ) ??
    null
  );
}

File: runtime/apps/orchestrator/lib/symphony_elixir/model_tiers.ex
defmodule SymphonyElixir.ModelTiers do
  @moduledoc """
  Runtime mirror of the platform model-tier registry.

  This file mirrors `contracts/model-tiers.ts` from the platform cutover
  contract. The registry assigns concrete provider/model pairs to execution
  adequacy tiers; `"any"` is only valid as a floor and is never returned as a
  model tier.
  """

  @type tier :: :frontier | :mid | :local
  @type floor :: tier() | :any

  @tier_order %{local: 0, mid: 1, frontier: 2}

  @registry %{
    {"anthropic", "claude-opus-4-7"} => :frontier,
    {"anthropic", "claude-sonnet-4-6"} => :frontier,
    {"anthropic", "claude-haiku-4-5"} => :mid,
    {"openai", "gpt-4.1"} => :frontier,
    {"openai", "gpt-4.1-mini"} => :mid,
    {"openai", "gpt-4o"} => :frontier,
    {"openai", "gpt-4o-mini"} => :mid,
    {"openai", "o3"} => :frontier,
    {"openai", "o3-mini"} => :mid,
    {"openai", "o1"} => :frontier,
    {"openai_codex", "gpt-4o"} => :frontier,
    {"openai_codex", "gpt-4.1"} => :frontier,
    {"openai_codex", "o3"} => :frontier,
    {"openai_compatible", "llama-3.1-405b-instruct"} => :mid,
    {"openai_compatible", "qwen2.5-coder-32b"} => :mid,
    {"google", "gemini-2.5-pro"} => :frontier,
    {"google", "gemini-2.5-flash"} => :mid,
    {"google", "gemini-2.0-flash"} => :mid,
    {"xai", "grok-4"} => :frontier,
    {"xai", "grok-3"} => :frontier,
    {"xai", "grok-3-mini"} => :mid,
    {"mistral", "mistral-large-2"} => :frontier,
    {"mistral", "mistral-medium"} => :mid,
    {"mistral", "codestral"} => :mid,
    {"groq", "llama-3.3-70b-versatile"} => :mid,
    {"groq", "llama-3.1-70b"} => :mid,
    {"groq", "mixtral-8x7b"} => :mid,
    {"openrouter", "anthropic/claude-opus-4-7"} => :frontier,
    {"openrouter", "anthropic/claude-sonnet-4-6"} => :frontier,
    {"openrouter", "openai/gpt-4o"} => :frontier,
    {"openrouter", "google/gemini-2.5-pro"} => :frontier,
    {"openrouter", "meta-llama/llama-3.1-405b"} => :frontier,
    {"together", "meta-llama/Llama-3.1-405B-Instruct"} => :frontier,
    {"together", "meta-llama/Llama-3.3-70B-Instruct"} => :mid,
    {"together", "mistralai/Mixtral-8x22B-Instruct"} => :mid,
    {"perplexity", "sonar-pro"} => :mid,
    {"perplexity", "sonar"} => :mid,
    {"perplexity", "llama-3.1-sonar-large-128k-online"} => :mid,
    {"azure", "gpt-4o"} => :frontier,
    {"azure", "gpt-4o-mini"} => :mid,
    {"azure", "o3"} => :frontier,
    {"bedrock", "anthropic.claude-opus-4-7-v1:0"} => :frontier,
    {"bedrock", "anthropic.claude-sonnet-4-6-v1:0"} => :frontier,
    {"bedrock", "meta.llama3-1-405b-instruct-v1:0"} => :frontier,
    {"bedrock", "mistral.mistral-large-2407-v1:0"} => :frontier,
    {"bedrock", "amazon.nova-pro-v1:0"} => :mid
  }

  @wildcards %{
    "openai_compatible" => :local
  }

  @spec tier_of(String.t() | atom() | nil, String.t() | atom() | nil) :: tier() | nil
  def tier_of(provider, model) do
    provider = normalize_string(provider)
    model = normalize_string(model)

    cond do
      is_nil(provider) or is_nil(model) ->
        nil

      tier = Map.get(@registry, {provider, model}) ->
        tier

      true ->
        Map.get(@wildcards, provider)
    end
  end

  @spec meets_floor?(tier() | nil, floor() | String.t() | atom() | nil) :: boolean()
  def meets_floor?(_tier, floor) when floor in [nil, "", :any, "any"], do: true

  def meets_floor?(nil, floor), do: meets_floor?(:local, floor)

  def meets_floor?(tier, floor) do
    with tier when tier in [:frontier, :mid, :local] <- normalize_tier(tier),
         floor when floor in [:frontier, :mid, :local] <- normalize_tier(floor) do
      Map.fetch!(@tier_order, tier) >= Map.fetch!(@tier_order, floor)
    else
      _ -> false
    end
  end

  @spec supported_floors() :: [String.t()]
  def supported_floors, do: ~w(any frontier mid local)

  defp normalize_tier(value) when value in [:frontier, :mid, :local, :any], do: value

  defp normalize_tier(value) when is_binary(value) do
    case normalize_string(value) do
      "frontier" -> :frontier
      "mid" -> :mid
      "local" -> :local
      "any" -> :any
      _ -> nil
    end
  end

  defp normalize_tier(_value), do: nil

  defp normalize_string(value) when is_binary(value) do
    value = String.trim(value)
    if value == "", do: nil, else: value
  end

  defp normalize_string(value) when is_atom(value) and not is_nil(value),
    do: value |> Atom.to_string() |> normalize_string()

  defp normalize_string(_value), do: nil
end
